### PR TITLE
TDirectory file space should never be released.

### PR DESCRIPTION
**Pull request by pcanal**
In https://root-forum.cern.ch/t/re-r-unzip-header-error/38463, the file contains a TDirectory and
a TTree with the same name.  Consequently the TKey for the TDirectory was deleted by AutoSave
leading to the corresponding portion of the file to be reused (for the 2nd basket of the branch ped_id) but TDirectory assumes that they key is never ever deleted and thus is rewriting ontop of the basket.

### CODE REVIEW

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
> The commit fixes a bug but it looks like we have no test for this case. This being TTree, I would warmly suggest a test is added for this case

Well ... that was of course useful ... as thinking about it more made it clear that there more similar cases.   See https://github.com/root-project/roottest/pull/487

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

File: tree/tree/src/TTree.cxx [hunk 1]
```diff
@@ -1494,7 +1494,7 @@ Long64_t TTree::AutoSave(Option_t* option)
       nbytes = fDirectory->WriteTObject(this,"","overwrite");
    } else {
       nbytes = fDirectory->WriteTObject(this); //nbytes will be 0 if Write failed (disk space exceeded)
-      if (nbytes && key) {
+      if (nbytes && key && strcmp("TTree", key->GetClassName()) == 0) {
```
Comment (jblomer) [inline tree/tree/src/TTree.cxx:1497]: Could this be a derived class name, e.g. `TNTuple`?

Comment (pcanal) [reply tree/tree/src/TTree.cxx:1497]: Good point !

**Comment by pcanal** [comment]
@phsft-bot build just on ROOT-performance-centos7-multicore/default with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build just on ROOT-performance-centos7-multicore/default with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by Axel-Naumann** [comment]
@pcanal please check travis clang-tidy-analyzer output; consider marking TKey final, or calling `TKey::DeleteBuffer()` (i.e. avoiding the vtable evaluation).

**Comment by pcanal** [comment]
> consider marking TKey final,

TBasket inherits from TKey.

**Comment by pcanal** [comment]
@phsft-bot build just on mac1015/cxx17 with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build just on mac1015/cxx17 with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
@phsft-bot build just on mac1015/cxx17 with flags -DCTEST_TEST_EXCLUDE_NONE=On

**Comment by pcanal** [comment]
The failures are seemingly unrelated.

**Pull request merged**